File: array-hash/560-Subarray-Sum-Equals-k.go
func subarraySum(nums []int, k int) int {
	count := 0
	prefixSum := 0
	countMap := make(map[int]int)
	countMap[0] = 1

	for _, num := range nums {
		prefixSum += num
		if val, found := countMap[prefixSum-k]; found {
			count += val
		}

		countMap[prefixSum]++
	}

	return count
}

File: reference/array-hash/560-Subarray-Sum-Equals-k.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        count = 0
        prefixSum = 0
        countMap = defaultdict(int)

        countMap[0] = 1

        for num in nums:
            prefixSum += num

            target = prefixSum - k
            if target in countMap:
                count += countMap[target]
            
            countMap[prefixSum] += 1
        return count